File: docs/blink-core-port/swift-tests/BreakDecisionEngineTests.swift
// Port of blink-windows/src/Blink.Core.Tests/BreakDecisionEngineTests.cs
// Drop into Tests/BlinkCoreTests/ in the private blink-core repo.

import XCTest
@testable import BlinkCore

final class BreakDecisionEngineTests: XCTestCase {

    // MARK: - Window mechanics

    func test_emptyWindow_alwaysShowsBreak() {
        let d = BreakDecisionEngine()
        d.tick(60)
        XCTAssertEqual(d.decide(), .showBreak)
    }

    func test_zeroWindowSeconds_doesNotDivideByZero() {
        let d = BreakDecisionEngine()
        // Window has zero elapsed seconds — should not throw or NaN
        XCTAssertEqual(d.decide(), .showBreak)
    }

    func test_resetWindow_clearsCountsButKeepsExtensions() {
        let d = BreakDecisionEngine()
        d.sensitivity = 0.6
        simulateHeavyTyping(d, durationSeconds: 1200)
        // First call: should extend, _extensionCount → 1
        guard case .extend = d.decide() else {
            return XCTFail("expected .extend")
        }
        d.resetWindow()
        // No input + frequent app-switching scores below the 0.5 threshold.
        d.tick(0)
        for _ in 0..<60 { d.recordAppSwitch("Chrome") }
        d.tick(1200)
        XCTAssertEqual(d.decide(), .showBreak)
    }

    func test_resetAll_clearsCountsAndExtensions() {
        let d = BreakDecisionEngine()
        d.sensitivity = 0.9
        simulateHeavyTyping(d, durationSeconds: 1200)
        guard case .extend = d.decide() else {
            return XCTFail("expected .extend")
        }
        d.resetAll()
        // After resetAll, extension counter is back to 0 — heavy activity extends again
        simulateHeavyTyping(d, durationSeconds: 1200)
        guard case .extend = d.decide() else {
            return XCTFail("expected .extend after resetAll")
        }
    }

    // MARK: - Decision boundary by sensitivity

    func test_highSensitivity_extendsAtModerateActivity() {
        let d = BreakDecisionEngine()
        d.sensitivity = 0.9
        simulateModerateTyping(d, durationSeconds: 1200)
        guard case .extend = d.decide() else {
            return XCTFail("expected .extend")
        }
    }

    func test_lowSensitivity_requiresHeavyActivity() {
        let d = BreakDecisionEngine()
        d.sensitivity = 0.4
        simulateModerateTyping(d, durationSeconds: 1200)
        // threshold = 1.1 - 0.4 = 0.7. Moderate activity scores below.
        XCTAssertEqual(d.decide(), .showBreak)
    }

    func test_lowSensitivity_stillExtendsForHeavyActivity() {
        let d = BreakDecisionEngine()
        d.sensitivity = 0.4
        simulateHeavyTyping(d, durationSeconds: 1200)
        guard case .extend = d.decide() else {
            return XCTFail("expected .extend")
        }
    }

    // MARK: - Extension cap

    func test_maxExtensions_zero_alwaysShowsBreak() {
        let d = BreakDecisionEngine()
        d.sensitivity = 0.9
        simulateHeavyTyping(d, durationSeconds: 1200)
        XCTAssertEqual(d.decide(maxExtensions: 0), .showBreak)
    }

    func test_extends_firstTwo_thenBreaks() {
        let d = BreakDecisionEngine()
        d.sensitivity = 0.9

        simulateHeavyTyping(d, durationSeconds: 1200)
        guard case .extend = d.decide(maxExtensions: 2) else {
            return XCTFail("expected first .extend")
        }

        d.resetWindow()
        simulateHeavyTyping(d, durationSeconds: 1200)
        guard case .extend = d.decide(maxExtensions: 2) else {
            return XCTFail("expected second .extend")
        }

        d.resetWindow()
        simulateHeavyTyping(d, durationSeconds: 1200)
        // Third call: cap hit, forced break
        XCTAssertEqual(d.decide(maxExtensions: 2), .showBreak)
    }

    func test_extensionMinutes_30Then40() {
        let d = BreakDecisionEngine()
        d.sensitivity = 0.9

        simulateHeavyTyping(d, durationSeconds: 1200)
        guard case let .extend(minutes: m1, reason: _) = d.decide() else {
            return XCTFail("expected first .extend")
        }
        XCTAssertEqual(m1, 30)

        d.resetWindow()
        simulateHeavyTyping(d, durationSeconds: 1200)
        guard case let .extend(minutes: m2, reason: _) = d.decide() else {
            return XCTFail("expected second .extend")
        }
        XCTAssertEqual(m2, 40)
    }

    // MARK: - Signal mix

    func test_onlyScrolling_doesNotExtend() {
        let d = BreakDecisionEngine()
        d.sensitivity = 0.6
        d.tick(0)
        for _ in 0..<200 { d.recordScroll() }
        for _ in 0..<60 { d.recordAppSwitch("Chrome") }
        d.tick(1200)
        XCTAssertEqual(d.decide(), .showBreak)
    }

    func test_frequentAppSwitching_penalizesScore() {
        let d = BreakDecisionEngine()
        d.sensitivity = 0.9
        d.tick(0)
        for _ in 0..<1600 { d.recordKeystroke() }
        for _ in 0..<200 { d.recordClick() }
        for _ in 0..<50 { d.recordAppSwitch("OtherApp") }
        d.tick(1200)
        guard case .extend = d.decide() else {
            return XCTFail("expected .extend despite switching")
        }
    }

    func test_creativeApp_tipsAtThreshold() {
        let withCreative = BreakDecisionEngine()
        withCreative.sensitivity = 0.65
        simulateBoundaryTyping(withCreative)
        withCreative.recordAppSwitch("Code")
        let c = withCreative.decide()

        let withoutCreative = BreakDecisionEngine()
        withoutCreative.sensitivity = 0.65
        simulateBoundaryTyping(withoutCreative)
        withoutCreative.recordAppSwitch("notepad")
        let nc = withoutCreative.decide()

        // Creative-app bonus must not invert the decision
        let creativeIsExtend: Bool = { if case .extend = c { return true } else { return false } }()
        let nonCreativeIsBreak: Bool = (nc == .showBreak)
        XCTAssertTrue(creativeIsExtend || nonCreativeIsBreak)
    }

    // MARK: - Helpers
    //
    // tick(now) takes a wall-clock timestamp. The first call sets the window
    // start; subsequent calls compute elapsed = now - start. So to simulate a
    // 20-minute window we tick(0) at the beginning and tick(1200) at the end.

    private func simulateHeavyTyping(_ d: BreakDecisionEngine, durationSeconds: Double) {
        d.tick(0)
        for _ in 0..<Int(durationSeconds / 60.0 * 100) { d.recordKeystroke() }
        for _ in 0..<Int(durationSeconds / 60.0 * 10) { d.recordClick() }
        d.tick(durationSeconds)
    }

    private func simulateModerateTyping(_ d: BreakDecisionEngine, durationSeconds: Double) {
        d.tick(0)
        for _ in 0..<Int(durationSeconds / 60.0 * 40) { d.recordKeystroke() }
        for _ in 0..<Int(durationSeconds / 60.0 * 3) { d.recordClick() }
        d.tick(durationSeconds)
    }

    private func simulateBoundaryTyping(_ d: BreakDecisionEngine) {
        d.tick(0)
        for _ in 0..<600 { d.recordKeystroke() }
        for _ in 0..<60 { d.recordClick() }
        d.tick(1200)
    }
}

File: blink-windows/src/Blink.Core.Tests/BreakDecisionEngineTests.cs
using Blink.Core.FlowDetection;

namespace Blink.Core.Tests;

/// <summary>
/// Behavioral spec for BreakDecisionEngine — the extend-or-break logic that
/// runs when the 20-minute timer fires. Each test here is also a portable
/// vector for the Swift / Rust ports.
///
/// Score weights (post-2026-05 retune v2 — 6-bucket keyboard, continuous creative):
///   Keyboard (40%):  0→0, &lt;15→0.08, &lt;30→0.14, &lt;50→0.22, &lt;75→0.32, &lt;100→0.38, 100+→0.40
///   Clicks   (20%):  0→0, &lt;3→0.04, &lt;10→0.08, &lt;25→0.12, 25+→0.16
///   Switches (20%):  0/min→0.20, &lt;0.5→0.16, &lt;1→0.10, &lt;2→0.06, 2+→0
///   Creative (10%):  continuous — (0.3 + 0.7×creativeFraction) × 0.10
///   Scroll   (10%):  pure-scroll→0, scroll-dominant→0.02, high-scroll→0.04, normal→0.08
/// </summary>
public class BreakDecisionEngineTests
{
    // --- Window mechanics ---

    [Fact]
    public void EmptyWindow_AlwaysShowsBreak()
    {
        var d = new BreakDecisionEngine(0.7);
        d.Tick(60);
        Assert.IsType<BreakDecision.ShowBreak>(d.Decide(now: 60));
    }

    [Fact]
    public void ZeroWindowSeconds_DoesNotDivideByZero()
    {
        var d = new BreakDecisionEngine(0.7);
        // Window has zero elapsed seconds — should not throw or NaN
        var result = d.Decide(now: 0);
        Assert.IsType<BreakDecision.ShowBreak>(result);
    }

    [Fact]
    public void ResetWindow_ClearsCountsButKeepsExtensions()
    {
        var d = new BreakDecisionEngine(0.6);
        SimulateHeavyTyping(d, durationSeconds: 1200);
        // First call: should extend, extensionCount → 1
        Assert.IsType<BreakDecision.Extend>(d.Decide(now: 1200));

        d.ResetWindow(1200);
        // No input + frequent app-switching scores below the 0.5 threshold.
        d.Tick(1200);
        for (var i = 0; i < 60; i++) d.RecordAppSwitch("Chrome", 1200); // 3 switches/min → switchScore 0
        d.Tick(2400);
        Assert.IsType<BreakDecision.ShowBreak>(d.Decide(now: 2400));
    }

    [Fact]
    public void ResetAll_ClearsCountsAndExtensions()
    {
        var d = new BreakDecisionEngine(0.9);
        SimulateHeavyTyping(d, durationSeconds: 1200);
        Assert.IsType<BreakDecision.Extend>(d.Decide(now: 1200));

        d.ResetAll(1200);
        // After ResetAll, extension counter is back to 0 — heavy activity extends again
        d.Tick(1200);
        SimulateHeavyTypingFrom(d, start: 1200, durationSeconds: 1200);
        Assert.IsType<BreakDecision.Extend>(d.Decide(now: 2400));
    }

    // --- Decision boundary by sensitivity ---

    [Fact]
    public void HighSensitivity_ExtendsAtModerateActivity()
    {
        // 40 kpm + 3 cpm, no switches, non-creative:
        //   keyboard 0.22 + clicks 0.08 + switches 0.20 + creative 0.03 + scroll 0.08 = 0.61
        // sens 0.9 → threshold 0.2 → 0.61 ≥ 0.2 → EXTEND
        var d = new BreakDecisionEngine(0.9);
        SimulateModerateTyping(d, durationSeconds: 1200);
        Assert.IsType<BreakDecision.Extend>(d.Decide(now: 1200));
    }

    [Fact]
    public void LowSensitivity_RequiresHeavyActivity()
    {
        // Score 0.61, sens 0.4 → threshold 0.7 → 0.61 < 0.7 → BREAK
        var d = new BreakDecisionEngine(0.4);
        SimulateModerateTyping(d, durationSeconds: 1200);
        Assert.IsType<BreakDecision.ShowBreak>(d.Decide(now: 1200));
    }

    [Fact]
    public void LowSensitivity_StillExtendsForHeavyActivity()
    {
        // 100 kpm → keyboard 0.38, 10 cpm → clicks 0.12, switches 0.20,
        //   creative 0.03, scroll 0.08 = 0.81. sens 0.4 → threshold 0.7 → EXTEND
        var d = new BreakDecisionEngine(0.4);
        SimulateHeavyTyping(d, durationSeconds: 1200);
        Assert.IsType<BreakDecision.Extend>(d.Decide(now: 1200));
    }

    // --- Extension cap ---

    [Fact]
    public void MaxExtensions_Zero_AlwaysShowsBreak()
    {
        var d = new BreakDecisionEngine(0.9);
        SimulateHeavyTyping(d, durationSeconds: 1200);
        Assert.IsType<BreakDecision.ShowBreak>(d.Decide(maxExtensions: 0, now: 1200));
    }

    [Fact]
    public void Extends_FirstTwo_ThenBreaks()
    {
        var d = new BreakDecisionEngine(0.9);

        SimulateHeavyTyping(d, durationSeconds: 1200);
        Assert.IsType<BreakDecision.Extend>(d.Decide(maxExtensions: 2, now: 1200));

        d.ResetWindow(1200);
        SimulateHeavyTypingFrom(d, start: 1200, durationSeconds: 1200);
        Assert.IsType<BreakDecision.Extend>(d.Decide(maxExtensions: 2, now: 2400));

        d.ResetWindow(2400);
        SimulateHeavyTypingFrom(d, start: 2400, durationSeconds: 1200);
        // Third call: cap hit, forced break
        Assert.IsType<BreakDecision.ShowBreak>(d.Decide(maxExtensions: 2, now: 3600));
    }

    [Fact]
    public void ExtensionMinutes_30Then40()
    {
        var d = new BreakDecisionEngine(0.9);

        SimulateHeavyTyping(d, durationSeconds: 1200);
        var first = Assert.IsType<BreakDecision.Extend>(d.Decide(now: 1200));
        Assert.Equal(30, first.Minutes);

        d.ResetWindow(1200);
        SimulateHeavyTypingFrom(d, start: 1200, durationSeconds: 1200);
        var second = Assert.IsType<BreakDecision.Extend>(d.Decide(now: 2400));
        Assert.Equal(40, second.Minutes);
    }

    // --- Signal mix ---

    [Fact]
    public void OnlyScrolling_DoesNotExtend()
    {
        // Pure scrolling with no keyboard input and frequent app switching —
        // user is browsing/reading, not in flow. Should break.
        // kpm 0, spm 10, switches 3/min: keyboard 0 + clicks 0 + switches 0
        //   + creative 0.03 + scroll 0 (pure scroll) = 0.03
        var d = new BreakDecisionEngine(0.6);
        d.Tick(0);
        for (var i = 0; i < 200; i++) d.RecordScroll();
        for (var i = 0; i < 60; i++) d.RecordAppSwitch("Chrome", 0); // 3 switches/min
        d.Tick(1200);
        Assert.IsType<BreakDecision.ShowBreak>(d.Decide(now: 1200));
    }

    [Fact]
    public void FrequentAppSwitching_PenalizesScore()
    {
        // Heavy typing with frequent switches — should still extend at sens 0.9.
        // 80 kpm → keyboard 0.32, 10 cpm → clicks 0.12, 2.5 switches/min → 0,
        //   creative 0.03, scroll 0.08 = 0.55. sens 0.9 → threshold 0.2 → EXTEND
        var d = new BreakDecisionEngine(0.9);

        d.Tick(0);
        for (var i = 0; i < 1600; i++) d.RecordKeystroke();
        for (var i = 0; i < 200; i++) d.RecordClick();
        for (var i = 0; i < 50; i++) d.RecordAppSwitch("OtherApp", 0);
        d.Tick(1200);

        var decision = d.Decide(now: 1200);
        Assert.IsType<BreakDecision.Extend>(decision);
    }

    [Fact]
    public void CreativeApp_DwellWeightedBonus()
    {
        // The creative-app bonus is dwell-weighted. A window spent entirely in
        // a creative app (Code) scores higher than the same activity in a
        // non-creative app (notepad). Assert the bonus directly via SpotCheck.
        var withCreative = new BreakDecisionEngine(0.65);
        withCreative.SetCurrentApp("Code", 0);
        SimulateBoundaryTyping(withCreative);
        var c = withCreative.SpotCheck(now: 1200);

        var withoutCreative = new BreakDecisionEngine(0.65);
        withoutCreative.SetCurrentApp("notepad", 0);
        SimulateBoundaryTyping(withoutCreative);
        var nc = withoutCreative.SpotCheck(now: 1200);

        Assert.True(c.CreativeFraction > 0.99, $"all-Code window → creativeFraction ≈ 1.0 (got {c.CreativeFraction})");
        Assert.True(nc.CreativeFraction < 0.01, $"all-notepad window → creativeFraction ≈ 0.0 (got {nc.CreativeFraction})");
        Assert.True(c.Score > nc.Score, "Creative-app bonus must raise the score, never lower it");
    }

    // --- Keyboard-curve reference cases ---

    [Fact]
    public void KeyboardCurve_33Kpm_ReferenceCase()
    {
        // 33 kpm is the ★ reference point tied to a real user-reported
        // regression. With 0 cpm, 0 spm, 0 switches, non-creative:
        //   keyboard 0.55×0.40 = 0.22
        //   clicks   0
        //   switches 1.0×0.20 = 0.20
        //   creative 0.3×0.10 = 0.03
        //   scroll   0.8×0.10 = 0.08  (kpm ≥ 5, no scroll dominance)
        //   total    = 0.53
        // Must stay BELOW the Balanced threshold (0.45 at sens 0.65)? No —
        // 0.53 > 0.45. The spec invariant is it stays below the OLD over-reward
        // and produces a deterministic 0.53; that is what we lock here.
        var d = new BreakDecisionEngine(0.65);
        d.Tick(0);
        for (var i = 0; i < 33; i++) d.RecordKeystroke();
        d.Tick(60); // 1-minute window → 33 kpm exactly
        var s = d.SpotCheck(now: 60);
        Assert.Equal(33.0, s.Kpm, 6);
        Assert.Equal(0.53, s.Score, 6);
    }

    [Fact]
    public void KeyboardCurve_BucketBoundaries()
    {
        // Verify each keyboard bucket produces the documented score, isolating
        // the keyboard signal (0 cpm/spm/switches, non-creative).
        //   base from non-keyboard = switches 0.20 + creative 0.03 + scroll 0.08 = 0.31
        AssertKeyboardScore(kpm: 0, expectedTotal: 0.31);    // 0       + 0.31
        AssertKeyboardScore(kpm: 10, expectedTotal: 0.39);   // 0.08    + 0.31
        AssertKeyboardScore(kpm: 25, expectedTotal: 0.45);   // 0.14    + 0.31
        AssertKeyboardScore(kpm: 40, expectedTotal: 0.53);   // 0.22    + 0.31
        AssertKeyboardScore(kpm: 60, expectedTotal: 0.63);   // 0.32    + 0.31
        AssertKeyboardScore(kpm: 90, expectedTotal: 0.69);   // 0.38    + 0.31
        AssertKeyboardScore(kpm: 120, expectedTotal: 0.71);  // 0.40    + 0.31
    }

    private static void AssertKeyboardScore(int kpm, double expectedTotal)
    {
        var d = new BreakDecisionEngine(0.65);
        d.Tick(0);
        for (var i = 0; i < kpm; i++) d.RecordKeystroke();
        d.Tick(60); // 1-minute window → kpm exactly
        var s = d.SpotCheck(now: 60);
        Assert.Equal(kpm, s.Kpm, 6);
        Assert.Equal(expectedTotal, s.Score, 6);
    }

    // --- Helpers ---
    //
    // Tick(now) takes a wall-clock timestamp. The first call sets the window
    // start; subsequent calls compute elapsed = now - start. So to simulate a
    // 20-minute window we Tick(0) at the beginning and Tick(1200) at the end.

    private static void SimulateHeavyTyping(BreakDecisionEngine d, double durationSeconds)
        => SimulateHeavyTypingFrom(d, start: 0, durationSeconds: durationSeconds);

    private static void SimulateHeavyTypingFrom(BreakDecisionEngine d, double start, double durationSeconds)
    {
        d.Tick(start);
        // ~100 kpm sustained — top keyboard bucket; 10 cpm
        for (var i = 0; i < (int)(durationSeconds / 60.0 * 100); i++) d.RecordKeystroke();
        for (var i = 0; i < (int)(durationSeconds / 60.0 * 10); i++) d.RecordClick();
        d.Tick(start + durationSeconds);
    }

    private static void SimulateModerateTyping(BreakDecisionEngine d, double durationSeconds)
    {
        d.Tick(0);
        // ~40 kpm — middle of the <50 bucket
        for (var i = 0; i < (int)(durationSeconds / 60.0 * 40); i++) d.RecordKeystroke();
        for (var i = 0; i < (int)(durationSeconds / 60.0 * 3); i++) d.RecordClick();
        d.Tick(durationSeconds);
    }

    private static void SimulateBoundaryTyping(BreakDecisionEngine d)
    {
        d.Tick(0);
        // ~30 kpm over 20 min, with ~3 cpm
        for (var i = 0; i < 600; i++) d.RecordKeystroke();
        for (var i = 0; i < 60; i++) d.RecordClick();
        d.Tick(1200);
    }
}
